### PR TITLE
Including El Attn optimization

### PR DESCRIPTION
**Pull request by NickNickGo**
Speed boost from 19.7 to 25.3.

### CODE REVIEW

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,1459 @@
+
+"""Apply the beam search optimizations to fairseq-v0.9.0"""
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:2]: Add copyright and modify this comment. (EL-attention related code?)

\#Copyright (c) Microsoft Corporation.
\#Licensed under the MIT License.

File: fastseq/optimizer/fairseq/generate.py [hunk 2]
```diff
@@ -60,7 +63,7 @@ def __init__(self, args, task, message_queue):
 
         self.args = args
         self.message_queue = message_queue
-        self.has_target = False
+        self.has_target = True
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/generate.py:66]: Why we need change this?

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,1459 @@
...
+        decoder (FairseqDecoder): the decoder
+    """
+
+    def transpose_enc_dec_kv_proj (self):
+        for i in range (0, len(self.decoder.layers)):
+            self.num_heads = self.decoder.layers[i].encoder_attn.num_heads
+            self.head_dim = self.decoder.layers[i].encoder_attn.head_dim
+
+
+            self.decoder.layers[i].encoder_attn.k_proj_weight_t = (
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:480]: Add a TODO. Keep the original k/v weight in CPU or remove them to avoid waste GPU memory.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,1459 @@
...
+            src_len = key.size(2)
+            tgt_len = 1
+            embed_dim = key.size(1)
+
+            torch.cuda.nvtx.range_push('Q reshape')
+            q = torch.addmm(self.q_proj.bias.view(1, -1),
+                    query.view(-1,embed_dim), self.q_proj.weight.T,
+                    beta=self.scaling, alpha=self.scaling)
+            q = (
+                q#.contiguous()
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:635]: Remove comment out code.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,1459 @@
...
+                    self.num_heads * self.head_dim)
+                .transpose(1, 2)
+                .contiguous()
+                ).cuda()
+            self.decoder.layers[i].encoder_attn.v_proj_bias_t = (
+                self.decoder.layers[i].encoder_attn.v_proj.bias
+                .view(1, 1, self.num_heads * self.head_dim)
+                ).cuda()
+
+@replace(TransformerModel, use_el_attn)
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:502]: Line 502 - 512 can be removed?

Comment (yuyan2do) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py:502]: Looks we cancel another modification here, so have to keep it. At least line 509 to 512 are doing nothing.

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 6]
```diff
@@ -62,7 +74,7 @@ def _reorder_encoder_out(self, encoder_out, new_order):
         return encoder_out
 
 
-@replace(TransformerModel)
+@replace(TransformerModel, not use_el_attn)
 class TransformerModelV2(TransformerModel):
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:78]: When use_el_attn is True, this V2 version won't be effect?

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,1459 @@
...
+            saved_state['prev_key'] = k.view(kv_bsz, self.num_heads, -1,
+                                             self.head_dim)
+            saved_state['prev_value'] = v.view(kv_bsz, self.num_heads, -1,
+                                               self.head_dim)
+            saved_state['prev_key_padding_mask'] = key_padding_mask
+            # In this branch incremental_state is never None
+            assert incremental_state is not None
+            self._set_input_buffer(incremental_state, saved_state)
+
+        #assert k is not None
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:719]: uncomment or remove

File: fastseq/optimizer/fairseq/generate.py [hunk 8]
```diff
@@ -19,9 +19,12 @@
 
 from fastseq.utils.api_decorator import replace
 
+
```
Comment (yuyan2do) [inline fastseq/optimizer/fairseq/generate.py:22]: Remove these new blank lines?

File: fastseq_cli/generate.py [hunk 9]
```diff
@@ -6,9 +6,34 @@
...
+from fairseq import options
+import os
+import argparse
 
-from fairseq_cli.generate import cli_main
+def parse_additional_args ():
+    parser = options.get_generation_parser()
+    parser.add_argument('--use-el-attn', action='store_true',
+            help='Use Efficient Lossless Attention optimization ? ')
+    parser.add_argument(
```
Comment (yuyan2do) [inline fastseq_cli/generate.py:17]: This is new code?

**Review by yuyan2do — APPROVED** [review]

File: tests/optimizer/fairseq/test_fairseq_optimizer.py [hunk 10]
```diff
@@ -64,11 +66,15 @@ def setUp(self):
         'lenpen': 2.0,
         'max_len_b': 140,
         'min_len': 55,
-        'no_repeat_ngram_size': 3
-    })
+        'no_repeat_ngram_size': 3,
+        #To test with el attention, please set 
+        #env variable USE_EL_ATTN=1
+        'use_el_attn':False,
```
Comment (yuyan2do) [inline tests/optimizer/fairseq/test_fairseq_optimizer.py:72]: Add a test for use_el_attn is True, and make sure it actually enabled.

File: examples/bart/README.md [hunk 11]
```diff
@@ -10,10 +10,11 @@ BART is sequence-to-sequence model trained with denoising as pretraining objecti
...
-  |     BatchSize    |       32      |        64       |      128       |      256*      |
-  |:----------------:|:-------------:|:---------------:|:--------------:|:--------------:|
-  | fairseq-0.9.0    | 2.4 samples/s |       OOM       |      OOM       |      OOM       |
-  | above + fastseq  | 8.1 samples/s | 13.3 samples/s  | 18.4 samples/s | 19.7 samples/s |
+  |     BatchSize    |       32      |        64       |      128       |      256*      |	320	  |
+  |:----------------:|:-------------:|:---------------:|:--------------:|:--------------:|:--------------:|
+  | fairseq-0.9.0    | 2.4 samples/s |       OOM       |      OOM       |      OOM       |      OOM       |
+  | above + fastseq  | 8.1 samples/s | 13.3 samples/s  | 18.4 samples/s | 19.7 samples/s |      OOM       |
+  | above + el_attn  | --- samples/s | ---- samples/s  | ---- samples/s | ---  samples/s | 25.3 samples/s |
 \* with `--max-tokens 131072` to avoid attn_weights' total number of elements exceed INT.MAX, which is a limitation for softmax op.
```
Comment (yuyan2do) [inline examples/bart/README.md:18]: Remove 256 setting which using trick (max-token).

File: benchmarks/benchmark_fs.sh [hunk 12]
```diff
@@ -110,7 +132,7 @@ for bs in "${bs_list[@]}"; do
             --max-len-b 140 \
             --no-repeat-ngram-size 3 \
             --lenpen 2.0 \
-            `#--print-alignment` \
+	    `#--print-alignment` \
```
Comment (feihugis) [inline benchmarks/benchmark_fs.sh:135]: minor: undo this change?

File: benchmarks/models/fs_bart.sh [hunk 13]
```diff
@@ -22,6 +22,12 @@ source utils.sh
     valid \
     32/64/128/256 \
     --max-tokens 131072
+./benchmark.sh \
+    fairseq+fastseq+el \
+    bart.large.cnn \
+    cnn_dm/len-1024.bin \
+    valid \
+    320
```
Comment (feihugis) [inline benchmarks/models/fs_bart.sh:30]: Try a few more batch_size, e.g., 128/256/320?

File: fastseq/config.py [hunk 14]
```diff
@@ -24,3 +24,7 @@
 # supported versions of fairseq
 MIN_FAIRSEQ_VERSION = '0.9.0'
 MAX_FAIRSEQ_VERSION = '0.9.0'
+
+#Set following variable to use Efficient-Lossless Attention 
+USE_EL_ATTN = os.getenv('USE_EL_ATTN', '0')
```
Comment (feihugis) [inline fastseq/config.py:29]: How about adding one more conversion here: '0' -> `False`, '1' -> `True` here? So that other places can directly use it as a boolean value.

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 15]
```diff
@@ -17,8 +18,19 @@
...
+logger = get_logger(__name__, logging.INFO)
+
+
+#Efficient-Lossless Attention
+use_el_attn = config.USE_EL_ATTN == '1'
+if use_el_attn:
+    logger.info(f"Using Efficient-Lossless Attention optimization")
+
+
+@replace(BeamSearch, True)
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:33]: It is not necessary here as 'True' is the default value. Remove it to be concise?

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 16]
```diff
@@ -47,7 +59,7 @@ def step(self, step, lprobs, scores):
         self.indices_buf.fmod_(vocab_size)
         return self.scores_buf, self.indices_buf, self.beams_buf
 
-@replace(TransformerEncoder)
+@replace(TransformerEncoder, not use_el_attn)
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:62]: How about using a global variable `USE_OPTIMIZED_CACHE_ATTN = not config.USE_EL_ATTN` instead of `not use_el_attn` to improve the readability.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 17]
```diff
@@ -0,0 +1,1459 @@
...
+from fairseq.sequence_generator import EnsembleModel
+from fairseq.tasks.fairseq_task import FairseqTask
+from fairseq.data.data_utils import collate_tokens
+from fastseq.utils.api_decorator import replace
+from fastseq.ops.ngram_repeat_block import NGramRepeatBlock
+from fastseq import config
+
+
+#Efficient-Lossless Attention
+use_el_attn = config.USE_EL_ATTN == '1'
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:32]: After converting `USE_EL_ATTN` to boolean type in config.py, `USE_EL_ATTN` can be directly used here.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 18]
```diff
@@ -0,0 +1,1459 @@
...
+
+
+@replace(collate_tokens, use_el_attn)
+def collate_tokens(values, pad_idx, eos_idx=None,
+        left_pad=False, move_eos_to_beginning=False):
+    """Convert a list of 1d tensors into a padded 2d tensor."""
+    size = max(v.size(0) for v in values)
+
+    pad_to_multiple = 8
+    if pad_to_multiple != 1 and size % pad_to_multiple != 0:
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:42]: `pad_to_multiple != 1` is always True because `pad_to_multiple` is hard coded as 8.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 1 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py]: Done

File: fastseq/optimizer/fairseq/generate.py (hunk 2 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/generate.py]: I changed for experimentation, reverted  back

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 3 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py:481]: deleted original weights after transposing.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 4 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py]: done

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 5 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py:507]: removed done

File: fastseq/optimizer/fairseq/beam_search_optimizer.py (hunk 6 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/beam_search_optimizer.py:78]: This class will not be replaced here , it will be replaced in el_attn_optimizer file.

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 7 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py]: done

File: fastseq/optimizer/fairseq/generate.py (hunk 8 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/generate.py]: done

File: fastseq_cli/generate.py (hunk 9 above)
Comment (NickNickGo) [reply fastseq_cli/generate.py:17]: yes , synched offline

File: tests/optimizer/fairseq/test_fairseq_optimizer.py (hunk 10 above)
Comment (NickNickGo) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py:72]: synched offline

File: examples/bart/README.md (hunk 11 above)
Comment (NickNickGo) [reply examples/bart/README.md:18]: done

File: benchmarks/models/fs_bart.sh (hunk 13 above)
Comment (NickNickGo) [reply benchmarks/models/fs_bart.sh:30]: EL attn basically helps in saving memory to that bs can be increased, think of it like cache optimization. So comparing same batch sizes is not really useful..

File: benchmarks/benchmark_fs.sh (hunk 12 above)
Comment (NickNickGo) [reply benchmarks/benchmark_fs.sh:135]: done

File: fastseq/config.py (hunk 14 above)
Comment (NickNickGo) [reply fastseq/config.py]: done

File: fastseq/optimizer/fairseq/beam_search_optimizer.py (hunk 15 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/beam_search_optimizer.py]: done

File: fastseq/optimizer/fairseq/beam_search_optimizer.py (hunk 16 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/beam_search_optimizer.py]: done

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 17 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py]: done

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 18 above)
Comment (NickNickGo) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py]: done

File: fastseq/models/__init__.py [hunk 19]
```diff
@@ -3,6 +3,7 @@
 """Register models"""
 
 import logging
+import sys
```
Comment (feihugis) [inline fastseq/models/__init__.py:6]: What's this for?

File: tests/optimizer/fairseq/test_fairseq_optimizer.py (hunk 10 above)
Comment (feihugis) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py:72]: What's the conclusion here?

Comment (NickNickGo) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py:72]: User will need to set this env variable before running unit test with EL attn.

File: fastseq/models/__init__.py (hunk 19 above)
Comment (NickNickGo) [reply fastseq/models/__init__.py:6]: This is required for exception at line 19 
https://github.com/microsoft/fastseq/blob/a21d906171869c813b28a72297ac6e00a3712979/fastseq/models/__init__.py#L19

File: tests/optimizer/fairseq/test_fairseq_optimizer.py (hunk 10 above)
Comment (feihugis) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py:72]: Then `config.USE_EL_ATTN` can be used here to avoid the above parameter `use_el_attn:False`, which needs to be manually changed in the above case.

Also update [here](https://github.com/microsoft/fastseq/blob/main/tests/run_fastseq_tests.sh#L6) by adding `USE_EL_ATTN=1 python ....` to trigger the tests for EL-Attention?

BTW, CI seems to be hung for a while.

File: tests/optimizer/fairseq/test_fairseq_optimizer.py [hunk 20]
```diff
@@ -83,6 +89,10 @@ def test_beam_search_optimizer(self, beam_size, batch_size, need_attn,
         """
         self.bart.model.make_generation_fast_(beamable_mm_beam_size=beam_size,
                                               need_attn=need_attn)
+        
+        if use_el_attn:
```
Comment (feihugis) [inline tests/optimizer/fairseq/test_fairseq_optimizer.py:93]: Use `config.USE_EL_ATTN`?

File: fastseq/models/__init__.py (hunk 19 above)
Comment (feihugis) [reply fastseq/models/__init__.py:6]: Gotcha. Thanks!

File: tests/optimizer/fairseq/test_fairseq_optimizer.py (hunk 10 above)
Comment (NickNickGo) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py]: EL attention is not supported for ProphetNet yet. Running all tests with USE_EL_ATTN=1 breaks regression.

File: tests/optimizer/fairseq/test_fairseq_optimizer.py (hunk 20 above)
Comment (NickNickGo) [reply tests/optimizer/fairseq/test_fairseq_optimizer.py]: done

**Review by feihugis — APPROVED** [review]
